### PR TITLE
`General`: Allow to restrict exercise import search to courses / exercises

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
# NOTE: Stacked PR based on https://github.com/ls1intum/Artemis/pull/5484

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
(https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

As requested by @krusche it would be helpful if we could search in the import dialog only for course exercises or only for exam exercises. 

### Description

Added new filter parameters and corresponding queries for the filter combination possibilities. 

### Steps for Testing

Test the search functionality in the import dialog of each exercise type: Text,Quiz,Modeling,Programming:
- CourseFilter active means course exercises should be searched for.
- ExamFilter active means exam exercises should be searched for.
- Activating both -> Both exam and course exercises are searched for
- Deactivating both is disabled as it would make no sense to do so from a user perspective

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
Test Coverage for critical methods:

| Class                      | Method               | Branch | Line |
|----------------------------|----------------------|--------|------|
| TextExerciseService        | getAllOnPageWithSize | 100%   | 100% |
| QuizExerciseService        | getAllOnPageWithSize | 100%   | 100% |
| ModelingExerciseService    | getAllOnPageWithSize | 100%   | 100% |
| ProgrammingExerciseService | getAllOnPageWithSize | 100%   | 100% |

### Screenshots

https://user-images.githubusercontent.com/29718932/181797266-b3c68cff-f20f-48c6-bb18-b7eb728e5f0f.mp4

### CODE REVIEW

**Review by jschmautzer — APPROVED** [review]
tested on ts1 - works as described

**Review by joaquinmurcia — APPROVED** [review]
Tested on TS1. Works as expected.

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
I don't like that you copied the code five times. Can this be avoided?

File: src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java [hunk 1]
```diff
@@ -740,6 +740,22 @@ void testInstructorGetResultsFromOwningCoursesNotEmpty() throws Exception {
...
+        exerciseIntegrationTestUtils.testCourseAndExamFilters("/api/text-exercises");
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = "ADMIN")
+    void testCourseAndExamFiltersAsAdmin() throws Exception {
+        database.addCourseWithOneReleasedTextExercise("Ankh");
+        database.addCourseExamExerciseGroupWithOneTextExercise("Ankh-Morpork");
+        exerciseIntegrationTestUtils.testCourseAndExamFilters("/api/text-exercises");
+    }
```
Comment (canberkanar) [inline src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java:757]: Please remove code duplication. My comment on https://github.com/ls1intum/Artemis/pull/5484 is valid here as well.

File: src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java [hunk 2]
```diff
@@ -847,6 +851,40 @@ private void testSearchTermMatchesId() throws Exception {
...
+
+        final Course course = database.addEmptyCourse();
+        final var now = ZonedDateTime.now();
+        QuizExercise exercise = ModelFactory.generateQuizExercise(now.minusDays(1), now.minusHours(2), QuizMode.INDIVIDUAL, course);
+        exercise.setTitle("Ankh");
+        quizExerciseRepository.save(exercise);
+
+        exerciseIntegrationTestUtils.testCourseAndExamFilters("/api/quiz-exercises/");
+    }
+
```
Comment (canberkanar) [inline src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java:887]: Please remove code duplication.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java [hunk 3]
```diff
@@ -185,6 +189,24 @@ private void testSearchTermMatchesId() throws Exception {
...
+
+    @Test
+    @WithMockUser(username = "admin", roles = "ADMIN")
+    void testCourseAndExamFiltersAsAdmin() throws Exception {
+        database.addCourseWithNamedProgrammingExerciseAndTestCases("Ankh");
+        database.addCourseExamExerciseGroupWithOneProgrammingExercise("Ankh-Morpork", "AnkhMorpork");
+
+        exerciseIntegrationTestUtils.testCourseAndExamFilters("/api/programming-exercises/");
+    }
+
```
Comment (canberkanar) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java:209]: Code duplication

**Review by canberkanar — CHANGES_REQUESTED** [review]
Please remove all code duplications. I have highlighted only "some" of them. My comment on https://github.com/ls1intum/Artemis/pull/5484 is valid here as well.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java (hunk 3 above)
Comment (stefanwaldhauser) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java:209]: > I don't like that you copied the code five times. Can this be avoided?

I see no possibility without the scope of this PR exploding.

**Comment by stefanwaldhauser** [comment]
> I don't like that you copied the code five times. Can this be avoided?

I think I can only do it for the test code to be honest without the scope of the PR completely exploding.

**Comment by Dominik-Weinzierl** [comment]
Yeah, there is an issue for that, but I see no reason why we should make it even worse :thinking: so @canberkanar and my point are still valid. So I suggest if you do not want to fix this in this PR, you should provide an additional PR refactoring the components and merge it first. Afterward, you can adapt your changes in this PR.

**Review by nairaarnez — DISMISSED** [review]
Tested in ts1, works as expected

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java (hunk 3 above)
Comment (stefanwaldhauser) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseServiceIntegrationTest.java]: @canberkanar I disagree in this case. Test with separate roles -> Separate Tests. Tests should only test one specific scenario

**Pull request merged**